### PR TITLE
GPS: Beidou should use B1, B1C as appropriate for concurrent Glonass

### PR DESCRIPTION
**Pull request by sensei-hacker**
Beidou has two options, B1 or B1C. B1 is faster to first fix, but the M10 cannot use B1 concurrently with Glonass.
Using B1C instead allows Beidou to be used concurrently with Glonass on M10.

This commit uses B1 when Beidou is active without Glonass, for best Beidou performance.
When Glonass is also active, it uses B1C to allow both Beidou and Glonass to be used together.

### Additional background information from u-blox datasheet:
From https://content.u-blox.com/sites/default/files/MAX-M10S_IntegrationManual_UBX-20053088.pdf

> 2.1.3.1 BeiDou B1I and B1C signals
BeiDou B1I and B1C signals differ in terms of the center frequency, bandwidth, and modulation.
Therefore, there are differences in performance and supported GNSS signals and features.
Benefits of using BeiDou B1I signal:
BeiDou B1I offers superior GNSS performance. High start-up sensitivity and fast time-to-first-fix
(TTFF) enable acquisition of a large number of BeiDou satellites. The tracking and reacquisition
sensitivity for acquired signals is approximately at the same level for BeiDou B1I and B1C.
• Faster TTFF and higher start-up sensitivity. BeiDou B1I signals are acquired significantly
faster and at a lower signal level than BeiDou B1C signals.
• Better availability. Higher start-up sensitivity results in a larger number of BeiDou satellites
tracked and used in navigation solution especially at low signal level.
• AssistNow Online support. Enhanced start-up sensitivity and TTFF.
• Power save mode (PSM) support. PSM cyclic tracking (PSMCT) and on/off (PSMOO) mode
operation are supported.
Benefits of using BeiDou B1C signal:
BeiDou B1C signal has the same center frequency as GPS L1 C/A enabling concurrent use of 4 GNSS
constellations. Multi-GNSS configurations with BeiDou B1C also have a lower power consumption
compared to those with BeiDou B1I.
• Concurrent reception of 4 GNSS with GPS L1 C/A, Galileo E1, BeiDou B1C, and GLONASS L1OF.
• Lower power consumption. No additional frequency band required for BeiDou B1C in multiGNSS constellations resulting in a lower power consumption during acquisition and tracking
phases.

### CODE REVIEW

**Comment by sensei-hacker** [comment]
Before commit:

gps_sats_before_commit
gps_before_commit

**Comment by sensei-hacker** [comment]
After commit:
gps_sats_after_commit
gps_before_after_commit

File: src/main/io/gps_ublox.c [hunk 1]
```diff
@@ -392,8 +392,9 @@ static void configureGNSS10(void)
 
             // Beidou
             {UBLOX_CFG_SIGNAL_BDS_ENA, gpsState.gpsConfig->ubloxUseBeidou},
-            {UBLOX_CFG_SIGNAL_BDS_B1_ENA, gpsState.gpsConfig->ubloxUseBeidou},
-            {UBLOX_CFG_SIGNAL_BDS_B1C_ENA, 0},
+            // Use B1 for Beidou without Glonass; B1C needed for concurrent use of Beidou and Glonass
+            {UBLOX_CFG_SIGNAL_BDS_B1_ENA,  gpsState.gpsConfig->ubloxUseBeidou &&  ! gpsState.gpsConfig->ubloxUseGlonass},
+            {UBLOX_CFG_SIGNAL_BDS_B1C_ENA, gpsState.gpsConfig->ubloxUseBeidou && gpsState.gpsConfig->ubloxUseGlonass},
```
Comment (mmosca) [inline src/main/io/gps_ublox.c]: What about my m9 that supports both? maybe add a check for hwversion != 10?

Comment (sensei-hacker) [reply src/main/io/gps_ublox.c]: The current check is:

if(gpsState.hwVersion >= UBX_HW_VERSION_UBLOX10 || (gpsState.swVersionMajor>=23 && gpsState.swVersionMinor >= 1)) {
            configureGNSS10()

Which does seem a bit odd to me - checking whether it's the cut down hardware OR has current firmware. Looks like you wrote that, Mosca, so perhaps you would know what's up with that check?

File: src/main/io/gps_ublox.c [hunk 2]
```diff
@@ -955,7 +956,7 @@ STATIC_PROTOTHREAD(gpsConfigure)
     // Configure GNSS for M8N and later
     if (gpsState.hwVersion >= UBX_HW_VERSION_UBLOX8) {
         gpsSetProtocolTimeout(GPS_SHORT_TIMEOUT);
-        if(gpsState.hwVersion >= UBX_HW_VERSION_UBLOX10 || (gpsState.swVersionMajor>=23 && gpsState.swVersionMinor >= 1)) {
+        if( (gpsState.swVersionMajor > 24) || (gpsState.swVersionMajor == 23 && gpsState.swVersionMinor >= 1) ) {
```
Comment (mmosca) [inline src/main/io/gps_ublox.c]: ```>= 24``` is probably what you want here

File: src/main/io/gps_ublox.c (hunk 1 above)
Comment (mmosca) [reply src/main/io/gps_ublox.c]: B1 works with  M9
Possibly even both B1 and B1C at the same time.

Maybe separating the Beidou options in BeidouB1 and BeidouB1C would be better, but that would need to happen in INAV 8.0

**Comment by mmosca** [comment]
So, at least on my m9 with SPG firmware, BDS_B1C is not supported.

image

**Comment by mmosca** [comment]
Unless we figure out a good way to identify supported channels, a good compromise is to expose channels in different settings.

The downside is the exploding number of settings that could end up as

**Comment by mmosca** [comment]
@sensei-hacker  are you targeting 7.1 or 8.0 for this?

**Comment by sensei-hacker** [comment]
> @sensei-hacker are you targeting 7.1 or 8.0 for this?

I've been working on this, trying M9 and M10 and various combinations.for the last couple days. There is a clear, simple fix which should probably go in 7.1 if possible. 

Later, for 8.0, maybe we can get more sophisticated with M9.  Within the next few hours I'll put in a PR against 7.1 that just fixes the clear and simple problem.

**Comment by sensei-hacker** [comment]
This is mostly superseded by #9646.  If future firmware updates from U-Blox create a problem on M9, I'll come back to this.  For now, #9646 fixes the issue on M10.

**Pull request closed without merging**